### ISSUE TITLE
Rust Language Support

### ISSUE THREAD

**Issue by XAMPPRocky** [EmbarkStudios/rust-gpu#78]
This is a tracking issue for adding support for using Rust language features and the `core` library with `rustc_codegen_spirv`.

## Language Support
- [x] Closures
- [ ] Enums
  - [x] Data-less enums. `enum Foo { A, B }`
  - [ ] Algebraic data types.  `enum Option<T> { Some(T), None }`
- [x] Loops
  - [x] `loop`
  - [x] `while`
  - [x] `for`
- [x] `if`
- [x] `match`
- [x] [Proc macros]
- [ ] Try (`?`) operator
- [ ] [128 bit integers](https://github.com/EmbarkStudios/rust-gpu/issues/196)
- [ ] [`union`](https://github.com/EmbarkStudios/rust-gpu/issues/415)
- [ ] Compiles on stable
  - Blocked on stabilisation of `-Z codegen-backend`, also see https://github.com/rust-lang/rust/issues/77933

[proc macros]: https://github.com/EmbarkStudios/rust-gpu/issues/192

## Compiler & Library Support
- [ ] Debug builds
  - Blocked on #81, #186 
- [ ] Iterators
- [ ] [`Option<T>`](https://github.com/EmbarkStudios/rust-gpu/issues/234)

## Features that don't work
- Dynamic Dispatch
- Pointer casts
- [Unsized slices](https://github.com/EmbarkStudios/rust-gpu/issues/378#issue-786964025) (`&[T]`)

## About Tracking Issues
Tracking issues are meant for providing status updates on the progress of a feature, it's not meant for bug reports or discussions on a feature. Please instead file a new issue or join the `#rust-gpu` channel on Discord.

**Comment by jarble** [comment]
Will closures be supported as well?

**Comment by XAMPPRocky** [comment]
@jarble Closures are currently supported.

**Comment by eddyb** [comment]
Should we update `README.md` and/or `docs` to indicate that e.g. `for` loops are now supported?

**Comment by mitchmindtree** [comment]
Hi folks! Is it possible that rust-gpu may one day support trait objects? I'm currently implementing ray tracing in one weekend with rust-gpu and there are a couple cases like `Hittable` and `Material`

**Comment by khyperia** [comment]
> Hi folks! Is it possible that rust-gpu may one day support trait objects?

I think at some point in the future, it may - however, we'd first have to get through a SPIR-V spec change to be able to support dynamic dispatch. The Slang folks are considering doing this (we've discussed a bit about what rust-gpu's needs would be if a SPIR-V feature for dynamic dispatch was proposed), but it's likely a ways off, and then even further beyond that to let drivers implement it, and beyond _that_ rust-gpu can start implementing it. So, I wouldn't hold your breath.

(`impl Trait` and friends all work 100% today, but yeah, unfortunately I don't think that's what you're asking)

**Comment by tema3210** [comment]
What about plain function pointers then? AFAIK spirv doesn't support these at all.

**Comment by khyperia** [comment]
> What about plain function pointers then? AFAIK spirv doesn't support these at all.

Yep, SPIR-V doesn't support them at all right now. With the Slang folks considering pushing through an implementation for dynamic dispatch in SPIR-V, plain function pointers may be able to piggyback on the same infrastructure as well, we'll see (probably more likely than not some hack would make it work). However, we haven't heard any updates in a while (and our unrelated requests for changes/clarifications to the SPIR-V spec have not made much progress at all), so I wouldn't hold your breath.

**Comment by ggaast** [comment]
Hey!
Is lifetimes supported in enums?
```rust
pub enum MyEnum<'a> {
    A {
        tmp: &'a Foo,
    },
    B {},
}
```

**Comment by tuguzT** [comment]
It seems that blockers of "debug builds" feature are all fixed at this moment.
I wonder if it is possible to implement such feature.
